src/ximea_ros_cam.cpp

```diff
@@ -869,9 +767,303 @@ void XimeaROSCam::triggerCb(const std_msgs::Empty::ConstPtr& msg) {
+    if(round_trip_duration > ros::Duration(0.0025)) // arbitary 2.5 ms (I see a 2 ms average with an xiC on my system)
+        ROS_WARN_STREAM_THROTTLE(3, "Camera timestamp query took " << round_trip_duration << " seconds." );    
+}
+
+ros::Time XimeaROSCam::iterpolateTimestamp(const XI_IMG& frame){
+    if(!camera_timestamp_supported_)  
+        return ros::Time::now();
+
+    sampleCameraTimestamp();
+
+    std::deque< std::pair<ros::Time,ros::Duration> >& queue = timestamp_queue_; // alias 
+    if(queue.size() < 2)
+    {  
+        ROS_ERROR("Not enough timestamp samples to interpolate.");
+        return ros::Time::now();
+    }
+
+    ros::Duration camera_stamp(frame.tsSec, frame.tsUSec * 1000);
+    // TODO adjust camera stamp to be the start of exposure (model dependant)
+    // consider using a precalculated bool to determin if exposure time should
+    // be deducted  
+    
+    bool clock_wrap = queue.front().second > queue.back().second; 
+    if(clock_wrap)
+    {
+        ROS_WARN("Clock wrap detected. Timestamp handling experimetal!");
+        // xiQ, xiD 
+        // 40 bit microsecond number - (overlaps after 305 hours)
+        // TimeStamp is recorded at the start of Data Readout 
+
+        // xiC, xiB, xiX, xiT 
+        // 64 bit 4-nanoseconds number (overlaps after 2339 years) (sic)
+        // TimeStamp is recorded at the start of Exposure.
+        // It is assumed that these cameras use the full tsSec value 
+
+        // untested experimental handling!
+        ros::Duration wrap_period;
+        ros::Duration rough_wrap_period = 
+                (queue.front().second - queue.back().second) + 
+                (queue.back().first - queue.front().first);
+
+        const uint64_t million = 1000000;
+        const uint64_t forty_bits = uint64_t(1) << 40;
+        const ros::Duration forty_bit_period(forty_bits / million, forty_bits % million);
+        const ros::Duration ros_duration_period(1<<31,0); // ros duration overflow
+
+        if((rough_wrap_period - forty_bit_period).sec == 0) 
+            wrap_period = forty_bit_period; // ~305 hours
+        else if((rough_wrap_period - ros_duration_period).sec == 0)  
+            wrap_period = ros::Duration(1<<31,0); // ~68 years (worlds longest running camera)
+        else
+            wrap_period = rough_wrap_period;
+         
+        ROS_WARN_STREAM("Adjusting timestamps using a period of " << wrap_period);
+        using iterator = std::deque< std::pair<ros::Time,ros::Duration> >::iterator;
+        for(iterator it = queue.begin(); it != queue.end(); ++it)
+            if(it->second > queue.back().second)
+                it->second -= wrap_period;
+        if(camera_stamp > queue.back().second)
+            camera_stamp -= wrap_period;    
+    }    
+
+    while(queue.size() > 2 && queue[1].second < camera_stamp)
+        queue.pop_front();
+
+    // interpolate stamp
+    double scalar = (camera_stamp - queue[0].second).toSec() /
```

Comment:
Be careful with this scalar computation, `(queue[1].second - queue[0].second).toSec()` can be fairly small and produce a large value. This may produce large jitters in the camera timestamp estimation.

I would recommend using higher order finite difference/differential approximation methods.